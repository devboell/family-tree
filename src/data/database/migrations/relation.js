export function up(knex, Promise) {
  return Promise.all([
    knex.schema.createTable('Relation', (table) => {
      table.increments('id')
      table.integer('person1Id')
      table.integer('person2Id')
      table.foreign('person1Id').references('Person.id')
      table.foreign('person2Id').references('Person.id')
    }),
  ])
}

export function down(knex, Promise) {
  return Promise.all([
    knex.schema.dropTable('Relation'),
  ])
}
