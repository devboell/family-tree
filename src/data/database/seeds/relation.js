const relations = `
  INSERT INTO 'Relation' VALUES
  (1,1,2),
  (2,4,5),
  (3,7,8),
  (4,9,10),
  (5,4,14),
  (6,15,16)
`

/*
const relations = `
  INSERT INTO 'Relation' VALUES
  (1,1,2)
`
*/

export function seed(knex) { // eslint-disable-line import/prefer-default-export
  return knex('Relation').truncate()
    .then(() =>
      knex.raw(relations),
    )
}
