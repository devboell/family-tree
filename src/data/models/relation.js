const knex = require('../connectors')


const findAll = () =>
  knex('Relation').select('*')

const findByPersonId = personId =>
  knex('Relation')
    .select('id')
    .where('person1Id', personId)
    .orWhere('person2Id', personId)

const remove = ids =>
  knex('Relation')
    .whereIn('id', ids)
    .del()

const create = (personId, partnerIds) => {
  const insertValues = partnerIds.map(id => ({ person1Id: personId, person2Id: id }))
  if (insertValues.length) {
    return knex('Relation').insert(insertValues)
  }
  return null
}

module.exports = {
  findAll,
  findByPersonId,
  create,
  remove,
}
