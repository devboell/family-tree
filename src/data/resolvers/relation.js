const find = require('lodash/fp/find')
const RelationModel = require('../models/relation')
const PersonModel = require('../models/person')

const relations = async () => {
  const allRelations = await RelationModel.findAll()
  const person1Ids = allRelations.map(rel => rel.person1Id)
  const person2Ids = allRelations.map(rel => rel.person2Id)
  const person1s = await PersonModel.findByIds(person1Ids)
  const person2s = await PersonModel.findByIds(person2Ids)

  const result = allRelations.map(rel => (
    {
      id: rel.id,
      person1: find({ id: rel.person1Id })(person1s),
      person2: find({ id: rel.person2Id })(person2s),
    }
  ))

  return result
}

module.exports = { relations }
