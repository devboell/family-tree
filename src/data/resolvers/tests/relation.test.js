import knex from '../../connectors'
import { relations } from '../relation'


describe('relation, resolvers', () => {
  beforeEach(async () => {
    await knex.migrate.rollback()
    await knex.migrate.latest()
    await knex.seed.run()
  })

  afterEach(async () => {
    await knex.migrate.rollback()
  })

  afterAll(() => {
    knex.destroy()
  })

  it('relations', async () => {
    expect(await relations()).toMatchSnapshot()
  })
})
