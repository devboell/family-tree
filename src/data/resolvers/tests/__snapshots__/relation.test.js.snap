// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relation, resolvers relations 1`] = `
Array [
  Object {
    "id": 1,
    "person1": Object {
      "birthDate": 1028,
      "bornToId": null,
      "deathDate": 1087,
      "gender": "M",
      "id": 1,
      "isMonarch": 1,
      "name": "William I",
      "title": "King of England",
    },
    "person2": Object {
      "birthDate": 1031,
      "bornToId": null,
      "deathDate": 1083,
      "gender": "F",
      "id": 2,
      "isMonarch": 0,
      "name": "Matilde of Flanders",
      "title": "Queen of England",
    },
  },
  Object {
    "id": 2,
    "person1": Object {
      "birthDate": 1068,
      "bornToId": 1,
      "deathDate": 1135,
      "gender": "M",
      "id": 4,
      "isMonarch": 1,
      "name": "Henry I",
      "title": "King of England",
    },
    "person2": Object {
      "birthDate": 1080,
      "bornToId": null,
      "deathDate": 1118,
      "gender": "F",
      "id": 5,
      "isMonarch": 0,
      "name": "Matilde of Scotland",
      "title": "Queen of England",
    },
  },
  Object {
    "id": 3,
    "person1": Object {
      "birthDate": 1102,
      "bornToId": 2,
      "deathDate": 1167,
      "gender": "F",
      "id": 7,
      "isMonarch": 0,
      "name": "Empress Matilde",
      "title": "Holy Roman Empress",
    },
    "person2": Object {
      "birthDate": 1113,
      "bornToId": null,
      "deathDate": 1151,
      "gender": "M",
      "id": 8,
      "isMonarch": 0,
      "name": "Geoffrey Plantagenet",
      "title": "Count of Anjou",
    },
  },
  Object {
    "id": 4,
    "person1": Object {
      "birthDate": 1133,
      "bornToId": 3,
      "deathDate": 1189,
      "gender": "M",
      "id": 9,
      "isMonarch": 1,
      "name": "Henry II",
      "title": "King of England",
    },
    "person2": Object {
      "birthDate": 1122,
      "bornToId": null,
      "deathDate": 1204,
      "gender": "F",
      "id": 10,
      "isMonarch": 0,
      "name": "Eleanor of Aquitane",
      "title": "Queen of England",
    },
  },
  Object {
    "id": 5,
    "person1": Object {
      "birthDate": 1068,
      "bornToId": 1,
      "deathDate": 1135,
      "gender": "M",
      "id": 4,
      "isMonarch": 1,
      "name": "Henry I",
      "title": "King of England",
    },
    "person2": Object {
      "birthDate": 1103,
      "bornToId": null,
      "deathDate": 1151,
      "gender": "F",
      "id": 14,
      "isMonarch": 0,
      "name": "Adeliza of Louvain",
      "title": "Queen of England",
    },
  },
  Object {
    "id": 6,
    "person1": Object {
      "birthDate": 1067,
      "bornToId": 1,
      "deathDate": 1137,
      "gender": "F",
      "id": 15,
      "isMonarch": 0,
      "name": "Adela of Normandy",
      "title": "Countess of Blois",
    },
    "person2": Object {
      "birthDate": 1045,
      "bornToId": null,
      "deathDate": 1102,
      "gender": "M",
      "id": 16,
      "isMonarch": 0,
      "name": "Stephen II Henry",
      "title": "Count of Blois",
    },
  },
]
`;
